### PR TITLE
!test Migrate multi node testkit to Netty 4.

### PR DESCRIPTION
**Pull request by He-Pin**
refs: https://github.com/apache/incubator-pekko/issues/462

I verified it locally with ` sbt cluster/MultiJvm/test`
previous: https://github.com/apache/incubator-pekko/pull/486

### CODE REVIEW

File: project/SbtMultiJvm.scala
Comment (He-Pin) [inline project/SbtMultiJvm.scala:380]: This is needed to get it works on linux and mac.

**Comment by pjfanning** [comment]
I agree that this can be merged to pekko main (when it is approved) and eventually released with pekko 1.1.0. 

I think that anyone who needs this code before the pekko 1.1.0 can easily create their own jar. It would probably be useful for a pekko-multi-node-netty4-testkit project to be created in GitHub for that project to release jars. The project can be archived after the pekko 1.1.0 release.

File: project/SbtMultiJvm.scala
Comment (pjfanning) [reply project/SbtMultiJvm.scala:380]: I don't like time based waits. On a slow machine, they can stop working. Our CI builds on VMs that can be very slow.

Is there any way that this code could be changed to ping the netty cluster and keep doing so until it gets notification that the cluster is ready?

**Review by pjfanning — CHANGES_REQUESTED** [review]

**Comment by He-Pin** [comment]
@pjfanning `ping` and `telnet` is not available on Windows, then I need to implement another `telnet` in java to test the controller is started up.

**Comment by pjfanning** [comment]
> @pjfanning `ping` and `telnet` is not available on Windows, then I need to implement another `telnet` in java to test the controller is started up.

When I say ping, I don't mean the Unix command. Wiktionary has this meaning (among many) of the word 'ping'.

A packet which a remote host is expected to echo, thus indicating its presence.

**Comment by He-Pin** [comment]
I will udpate the pr wit a `java` ping then.

File: project/SbtMultiJvm.scala
Comment (He-Pin) [inline project/SbtMultiJvm.scala:410]: @pjfanning I think it can be running on a slower box now.

Comment (pjfanning) [reply project/SbtMultiJvm.scala:410]: seems like a good solution

File: cluster/src/multi-jvm/resources/logback-test.xml
Comment (mdedetrich) [inline cluster/src/multi-jvm/resources/logback-test.xml:11]: Is this something that was needed just because of netty4 or is it generally helpful? If it is generally helpful I would do this change in a separate PR so that we can backport it to Pekko 1.0.x

File: multi-node-testkit/src/main/scala/org/apache/pekko/remote/testconductor/RemoteConnection.scala
Comment (mdedetrich) [inline multi-node-testkit/src/main/scala/org/apache/pekko/remote/testconductor/RemoteConnection.scala:140]: Is it worth it to make these shutdown gracefully constants configurable in typesafe config? As I understand these values are the underlying reasons why the implementation initially didn't work on Mac/Linux and if so I think it makes sense for users to be able to configure this without needing a new release incase there are further issues in the future.

File: cluster/src/multi-jvm/resources/logback-test.xml
Comment (mdedetrich) [inline cluster/src/multi-jvm/resources/logback-test.xml:11]: Is this something that was needed just because of netty4 or is it generally helpful? If it is generally helpful I would do this change in a separate PR so that we can backport it to Pekko 1.0.x

File: multi-node-testkit/src/main/scala/org/apache/pekko/remote/testconductor/RemoteConnection.scala
Comment (mdedetrich) [inline multi-node-testkit/src/main/scala/org/apache/pekko/remote/testconductor/RemoteConnection.scala:168]: Same as previous comment about configuration

File: project/SbtMultiJvm.scala
Comment (mdedetrich) [inline project/SbtMultiJvm.scala:174]: Although this change was needed for this feature specifically, I think it would be wiser to make this change in a separate PR because in general we should concatenate all `LICENSE` statements.

File: project/SbtMultiJvm.scala
Comment (mdedetrich) [inline project/SbtMultiJvm.scala:382]: As with previous comments about magic constants should this be made configurable (although in this specific case since we are dealing with test classes in sbt project build it should be done by cli args as is done elsewhere).

File: project/SbtMultiJvm.scala
Comment (mdedetrich) [inline project/SbtMultiJvm.scala:406]: Same with previous comment about constants.

**Review by mdedetrich — CHANGES_REQUESTED** [review]
Great work, PR generally looks fine. My main remarks are regarding magic constants and some changes I think be split out and merged on their own so they can be backported.

File: multi-node-testkit/src/main/scala/org/apache/pekko/remote/testconductor/RemoteConnection.scala
Comment (He-Pin) [reply multi-node-testkit/src/main/scala/org/apache/pekko/remote/testconductor/RemoteConnection.scala:140]: Noop, The reason it not working because of the `Player` not wanting the `Controller` started in the sbtMultiJvm. the old `shutdown` method is deprecated but still can be used.

**Comment by mdedetrich** [comment]
I have just run the cluster tests on the latest version of this PR and I can confirm that they all pass, good work! After the comments are addressed the PR looks to be in good shape to me.

I have attached the test log for those curios

[multi-jvm-test-main-6.txt](https://github.com/apache/incubator-pekko/files/12269096/multi-jvm-test-main-6.txt)

File: cluster/src/multi-jvm/resources/logback-test.xml
Comment (He-Pin) [reply cluster/src/multi-jvm/resources/logback-test.xml:11]: It because there are so much netty log when you run testing.

File: project/SbtMultiJvm.scala
Comment (He-Pin) [reply project/SbtMultiJvm.scala:382]: It's configurable with the `-Dmultinode.server-port`.

File: cluster/src/multi-jvm/resources/logback-test.xml
Comment (mdedetrich) [reply cluster/src/multi-jvm/resources/logback-test.xml:11]: I thought so, do you mind making this change in a separate PR so it can be back-ported separately to Pekko 1.0.x?

I can also do this, and technically speaking you shouldn't even need to update the PR since it would get resolve via a merge.

File: project/SbtMultiJvm.scala
Comment (mdedetrich) [reply project/SbtMultiJvm.scala:406]: I will resolve this for now, I don't want to block the PR because of this and if people have issues we can always update this later (its just tests)

File: cluster/src/multi-jvm/resources/logback-test.xml
Comment (mdedetrich) [reply cluster/src/multi-jvm/resources/logback-test.xml:11]: There was a miscommunication, I realize that `PoolThreadCache` and `Recycler` classes only exist in Netty 4 and not Netty 3 (just checked now), resolving conversation.

**Review by mdedetrich — APPROVED** [review]
So I am approving changes, previous comments are resolved. I am not fully aware of this part of the code so it would be good to get some more eyes before merging (we aren't in a massive rush to merge this right now).

File: project/SbtMultiJvm.scala
Comment (mdedetrich) [reply project/SbtMultiJvm.scala]: Regarding this point, maybe it makes sense to make this a separate issue to look at after this PR gets merged? The reason I am stating this is because the best way to find out if these time based constants are too low is by seeing how often the tests fail in our nightly github CI action and we can then iterate from there.

Of course most ideal would be to ping the netty cluster until its ready but I don't feel its valid to block the PR over this, for me its critical to get people to test this directly (via snpashots/building locally), hopefully in production to see if there aren't any regressions well before a release happens.

File: multi-node-testkit/src/main/scala/org/apache/pekko/remote/testconductor/RemoteConnection.scala
Comment (pjfanning) [reply multi-node-testkit/src/main/scala/org/apache/pekko/remote/testconductor/RemoteConnection.scala:168]: I agree with @mdedetrich - this would be better as a setting in reference.conf

File: project/SbtMultiJvm.scala
Comment (He-Pin) [reply project/SbtMultiJvm.scala:406]: I think that's ok, as it's just in tests and no one will touch those again, otherwise we need another：
```
multinode.max-time-waiting-to-connect-in-seconds=20 
```
like thing

File: project/SbtMultiJvm.scala
Comment (He-Pin) [reply project/SbtMultiJvm.scala:174]: Should I seperate this to a dedicated PR? but seems like only Netty 4 contains this thing.

File: multi-node-testkit/src/main/scala/org/apache/pekko/remote/testconductor/RemoteConnection.scala
Comment (He-Pin) [reply multi-node-testkit/src/main/scala/org/apache/pekko/remote/testconductor/RemoteConnection.scala]: I just updated the code without this, but which may waiting max to 15 seconds. the old code is deprecated.

File: project/SbtMultiJvm.scala
Comment (He-Pin) [inline project/SbtMultiJvm.scala:380]: @mdedetrich This is the real fix which can be a sleep(2000) or like thing. the error was an `CONNECT(2)` from the networking , which I was thing caused by resolver or my code bug.

File: project/SbtMultiJvm.scala
Comment (mdedetrich) [reply project/SbtMultiJvm.scala:174]: Yes please make a separate PR for this. Although it's true that this only effects netty 4 for now, in general of there multiple `LICENSE`'s we do want to concat them (i.e. this isn't just specific to netty 4)

File: multi-node-testkit/src/main/scala/org/apache/pekko/remote/testconductor/RemoteConnection.scala
Comment (He-Pin) [reply multi-node-testkit/src/main/scala/org/apache/pekko/remote/testconductor/RemoteConnection.scala]: @mdedetrich I updated it to `.shutdownGracefully(0L, 0L, TimeUnit.SECONDS)` which alian the old behavior.

**Review by pjfanning — APPROVED** [review]
lgtm

**Comment by He-Pin** [comment]
I run several times on gitpod. it works, I will do a final validation before merge this.

**Comment by He-Pin** [comment]
I ran it several times again and it do passed, so it would be good to go.

**Pull request merged**
